circuits/plonk/src/gate.rs

```diff
@@ -0,0 +1,65 @@
+/*****************************************************************************************************************
+
+This source file implements Plonk computation wire index primitive.
+
+*****************************************************************************************************************/
+
+use algebra::Field;
+
+#[derive(Clone)]
+pub struct CircuitGate<F: Field>
+{
+    pub l: (usize, usize),   // left input wire index and its permutation
```

Comment:
Could you improve the comment here? It's not clear to me what the meaning of the `(usize, usize)` pair is. Don't we only need a single `usize` per each of `l, r, o` saying e.g., the equivalence class of variable that it belongs to?

Reply:
The second part of the tuple, defines the index of the wire permutation. Permutation sets (or equivalence classes) have to be defined in some way. Defining a permutation as decomposition of the set of wires into the disjoint permutation sets is equivalent to defining a bijective map of a set of wires onto itself. Thus, defining where each wire goes under the permutation. The left part of the tuple is the wire index, the right part the index of the "permuted wire" index. This mapping is then used, upon the creation of the ConstraintSystem struct, to build the permutation polynomials. Another  way would be to define the permutation polys directly but it is possibly more elegant to define everything, including the permutation map, in the set of the gates that completely determine the constraint system. Will enhance the comment,